File: tna/transitions.py
"""Transition computation algorithms for TNA package."""

from __future__ import annotations

import numpy as np
import pandas as pd

from .utils import ensure_matrix, row_normalize


def compute_transitions(
    data: np.ndarray,
    states: list[str],
    type_: str = "relative",
    params: dict | None = None
) -> tuple[np.ndarray, np.ndarray]:
    """Compute transition matrix and initial probabilities from sequence data.

    Parameters
    ----------
    data : np.ndarray
        Sequence data (rows are sequences, columns are time steps)
    states : list of str
        List of state labels
    type_ : str
        Type of transitions to compute:
        - 'relative': Row-normalized transition probabilities
        - 'frequency': Raw transition counts
        - 'co-occurrence': Bidirectional co-occurrence
        - 'reverse': Reverse order transitions
        - 'n-gram': Higher-order n-gram transitions
        - 'gap': Non-adjacent transitions weighted by gap
        - 'window': Sliding window transitions
        - 'attention': Exponential decay weighted
    params : dict, optional
        Additional parameters for specific types:
        - n_gram: {'n': int} - order of n-gram (default: 2)
        - gap: {'max_gap': int, 'decay': float} - max gap and decay factor
        - window: {'size': int} - window size
        - attention: {'beta': float} - decay parameter

    Returns
    -------
    tuple
        (transition_matrix, initial_probabilities)
    """
    params = params or {}
    n_states = len(states)
    state_to_idx = {s: i for i, s in enumerate(states)}

    if type_ == "relative":
        weights, inits = _transitions_relative(data, state_to_idx, n_states)
    elif type_ == "frequency":
        weights, inits = _transitions_frequency(data, state_to_idx, n_states)
    elif type_ == "co-occurrence":
        weights, inits = _transitions_cooccurrence(data, state_to_idx, n_states)
    elif type_ == "reverse":
        weights, inits = _transitions_reverse(data, state_to_idx, n_states)
    elif type_ == "n-gram":
        n = params.get('n', 2)
        weights, inits = _transitions_ngram(data, state_to_idx, n_states, n)
    elif type_ == "gap":
        max_gap = params.get('max_gap', 5)
        decay = params.get('decay', 0.5)
        weights, inits = _transitions_gap(data, state_to_idx, n_states, max_gap, decay)
    elif type_ == "window":
        size = params.get('size', 3)
        weights, inits = _transitions_window(data, state_to_idx, n_states, size)
    elif type_ == "attention":
        beta = params.get('beta', 0.1)
        weights, inits = _transitions_attention(data, state_to_idx, n_states, beta)
    else:
        raise ValueError(f"Unknown transition type: {type_}")

    return weights, inits


def _get_valid_transitions(row: np.ndarray) -> list[tuple[int, str]]:
    """Get list of (position, state) for non-NA values in a row."""
    result = []
    for i, val in enumerate(row):
        if val is not None and not (isinstance(val, float) and np.isnan(val)):
            result.append((i, str(val)))
    return result


def _transitions_relative(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int
) -> tuple[np.ndarray, np.ndarray]:
    """Compute relative (row-normalized) transition probabilities."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count transitions
        for i in range(len(valid) - 1):
            from_state = valid[i][1]
            to_state = valid[i + 1][1]
            if from_state in state_to_idx and to_state in state_to_idx:
                counts[state_to_idx[from_state], state_to_idx[to_state]] += 1

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _transitions_frequency(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int
) -> tuple[np.ndarray, np.ndarray]:
    """Compute raw transition counts (frequency model)."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count transitions
        for i in range(len(valid) - 1):
            from_state = valid[i][1]
            to_state = valid[i + 1][1]
            if from_state in state_to_idx and to_state in state_to_idx:
                counts[state_to_idx[from_state], state_to_idx[to_state]] += 1

    # Normalize inits only
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return counts, inits


def _transitions_cooccurrence(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int
) -> tuple[np.ndarray, np.ndarray]:
    """Compute bidirectional co-occurrence matrix."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count co-occurrences (bidirectional adjacent pairs)
        for i in range(len(valid) - 1):
            state1 = valid[i][1]
            state2 = valid[i + 1][1]
            if state1 in state_to_idx and state2 in state_to_idx:
                idx1, idx2 = state_to_idx[state1], state_to_idx[state2]
                counts[idx1, idx2] += 1
                counts[idx2, idx1] += 1

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _transitions_reverse(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int
) -> tuple[np.ndarray, np.ndarray]:
    """Compute transitions in reverse order."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state (last in original sequence)
        last_state = valid[-1][1]
        if last_state in state_to_idx:
            inits[state_to_idx[last_state]] += 1

        # Count reverse transitions
        for i in range(len(valid) - 1, 0, -1):
            from_state = valid[i][1]
            to_state = valid[i - 1][1]
            if from_state in state_to_idx and to_state in state_to_idx:
                counts[state_to_idx[from_state], state_to_idx[to_state]] += 1

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _transitions_ngram(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int,
    n: int = 2
) -> tuple[np.ndarray, np.ndarray]:
    """Compute n-gram transitions (skip n-1 steps)."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count n-gram transitions (pairs separated by n-1 positions)
        for i in range(len(valid) - n + 1):
            from_state = valid[i][1]
            to_state = valid[i + n - 1][1]
            if from_state in state_to_idx and to_state in state_to_idx:
                counts[state_to_idx[from_state], state_to_idx[to_state]] += 1

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _transitions_gap(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int,
    max_gap: int = 5,
    decay: float = 0.5
) -> tuple[np.ndarray, np.ndarray]:
    """Compute transitions with gap-based decay weighting."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count weighted transitions for all gaps up to max_gap
        for i in range(len(valid)):
            from_state = valid[i][1]
            if from_state not in state_to_idx:
                continue
            for j in range(i + 1, min(i + max_gap + 1, len(valid))):
                to_state = valid[j][1]
                if to_state not in state_to_idx:
                    continue
                gap = j - i
                weight = decay ** (gap - 1)  # gap=1 has weight 1
                counts[state_to_idx[from_state], state_to_idx[to_state]] += weight

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _transitions_window(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int,
    size: int = 3
) -> tuple[np.ndarray, np.ndarray]:
    """Compute transitions within sliding windows."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count all pairs within each window
        for window_start in range(len(valid) - size + 1):
            window = valid[window_start:window_start + size]
            for i in range(len(window)):
                for j in range(i + 1, len(window)):
                    state1 = window[i][1]
                    state2 = window[j][1]
                    if state1 in state_to_idx and state2 in state_to_idx:
                        counts[state_to_idx[state1], state_to_idx[state2]] += 1

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _transitions_attention(
    data: np.ndarray,
    state_to_idx: dict,
    n_states: int,
    beta: float = 0.1
) -> tuple[np.ndarray, np.ndarray]:
    """Compute attention-weighted transitions with exponential decay."""
    counts = np.zeros((n_states, n_states))
    inits = np.zeros(n_states)

    for row in data:
        valid = _get_valid_transitions(row)
        if len(valid) == 0:
            continue

        # Count initial state
        first_state = valid[0][1]
        if first_state in state_to_idx:
            inits[state_to_idx[first_state]] += 1

        # Count attention-weighted transitions
        for i in range(len(valid)):
            from_state = valid[i][1]
            if from_state not in state_to_idx:
                continue
            for j in range(i + 1, len(valid)):
                to_state = valid[j][1]
                if to_state not in state_to_idx:
                    continue
                distance = j - i
                weight = np.exp(-beta * distance)
                counts[state_to_idx[from_state], state_to_idx[to_state]] += weight

    # Normalize
    weights = row_normalize(counts)
    inits = inits / inits.sum() if inits.sum() > 0 else inits

    return weights, inits


def _is_na(val) -> bool:
    """Check if a value is NA/None/NaN."""
    if val is None:
        return True
    if isinstance(val, float) and np.isnan(val):
        return True
    try:
        if pd.isna(val):
            return True
    except (TypeError, ValueError):
        pass
    return False


def compute_transitions_3d(
    data: np.ndarray,
    states: list[str],
    type_: str = "relative",
    params: dict | None = None
) -> np.ndarray:
    """Compute per-sequence transition counts as a 3D array.

    Matches R TNA's compute_transitions function exactly.
    Returns array of shape (n_sequences, n_states, n_states) where
    trans[k, i, j] = number of i->j transitions in sequence k.

    Parameters
    ----------
    data : np.ndarray
        Sequence data (rows are sequences, columns are time steps)
    states : list of str
        List of state labels
    type_ : str
        Transition type ('relative', 'frequency', 'co-occurrence', 'reverse')
    params : dict, optional
        Additional parameters

    Returns
    -------
    np.ndarray
        3D array of shape (n_sequences, n_states, n_states)
    """
    params = params or {}
    n_sequences = data.shape[0]
    n_steps = data.shape[1]
    n_states = len(states)
    state_to_idx = {s: i for i, s in enumerate(states)}

    trans = np.zeros((n_sequences, n_states, n_states))

    if type_ in ("relative", "frequency"):
        for col in range(n_steps - 1):
            for row in range(n_sequences):
                from_val = data[row, col]
                to_val = data[row, col + 1]
                if _is_na(from_val) or _is_na(to_val):
                    continue
                from_str = str(from_val)
                to_str = str(to_val)
                if from_str in state_to_idx and to_str in state_to_idx:
                    trans[row, state_to_idx[from_str], state_to_idx[to_str]] += 1

    elif type_ == "reverse":
        for col in range(n_steps - 1):
            for row in range(n_sequences):
                from_val = data[row, col + 1]
                to_val = data[row, col]
                if _is_na(from_val) or _is_na(to_val):
                    continue
                from_str = str(from_val)
                to_str = str(to_val)
                if from_str in state_to_idx and to_str in state_to_idx:
                    trans[row, state_to_idx[from_str], state_to_idx[to_str]] += 1

    elif type_ == "co-occurrence":
        for i in range(n_steps - 1):
            for j in range(i + 1, n_steps):
                for row in range(n_sequences):
                    from_val = data[row, i]
                    to_val = data[row, j]
                    if _is_na(from_val) or _is_na(to_val):
                        continue
                    from_str = str(from_val)
                    to_str = str(to_val)
                    if from_str in state_to_idx and to_str in state_to_idx:
                        fi = state_to_idx[from_str]
                        ti = state_to_idx[to_str]
                        trans[row, fi, ti] += 1
                        trans[row, ti, fi] += 1

    return trans


def compute_weights_from_3d(
    transitions: np.ndarray,
    type_: str = "relative",
    scaling: str | list[str] | None = None,
) -> np.ndarray:
    """Compute weight matrix from 3D per-sequence transitions.

    Matches R TNA's compute_weights + scale_weights:
        weights <- apply(transitions, c(2, 3), sum)
        scale_weights(weights, type, scaling, a)

    Parameters
    ----------
    transitions : np.ndarray
        3D array (n_sequences, n_states, n_states)
    type_ : str
        Model type ('relative' for row-normalization)
    scaling : str or list, optional
        Additional scaling to apply

    Returns
    -------
    np.ndarray
        2D weight matrix (n_states, n_states)
    """
    # apply(transitions, c(2, 3), sum) - sum over sequences
    weights = transitions.sum(axis=0)

    # scale_weights: row-normalize for "relative" type
    if type_ == "relative":
        weights = row_normalize(weights)

    # Apply additional scaling if requested
    if scaling:
        from .utils import apply_scaling as _apply_scaling
        weights, _ = _apply_scaling(weights, scaling)

    return weights


def compute_weights_from_matrix(
    mat: np.ndarray,
    type_: str = "relative"
) -> np.ndarray:
    """Process an existing weight/count matrix.

    Parameters
    ----------
    mat : np.ndarray
        Input matrix (counts or weights)
    type_ : str
        How to process:
        - 'relative': Row-normalize
        - 'frequency': Keep as-is

    Returns
    -------
    np.ndarray
        Processed weight matrix
    """
    if type_ == "relative":
        return row_normalize(mat)
    return mat.copy()

File: tna-js/src/core/transitions.ts
/**
 * Transition computation algorithms.
 * Port of Python tna/transitions.py
 */
import { Matrix, rowNormalize, applyScaling } from './matrix.js';
import type { ModelType, SequenceData, TransitionParams } from './types.js';

/** Check if a value is null/undefined/NaN. */
function isNA(val: string | null | undefined): boolean {
  return val === null || val === undefined || val === '';
}

/** Get list of (position, state) for non-NA values in a sequence row. */
function getValidTransitions(row: (string | null)[]): { pos: number; state: string }[] {
  const result: { pos: number; state: string }[] = [];
  for (let i = 0; i < row.length; i++) {
    const val = row[i];
    if (!isNA(val)) {
      result.push({ pos: i, state: val! });
    }
  }
  return result;
}

/**
 * Compute transition matrix and initial probabilities from sequence data.
 */
export function computeTransitions(
  data: SequenceData,
  states: string[],
  type: ModelType = 'relative',
  params?: TransitionParams,
): { weights: Matrix; inits: Float64Array } {
  const nStates = states.length;
  const stateToIdx = new Map<string, number>();
  states.forEach((s, i) => stateToIdx.set(s, i));

  switch (type) {
    case 'relative':
      return transitionsRelative(data, stateToIdx, nStates);
    case 'frequency':
      return transitionsFrequency(data, stateToIdx, nStates);
    case 'co-occurrence':
      return transitionsCooccurrence(data, stateToIdx, nStates);
    case 'reverse':
      return transitionsReverse(data, stateToIdx, nStates);
    case 'n-gram':
      return transitionsNgram(data, stateToIdx, nStates, params?.n ?? 2);
    case 'gap':
      return transitionsGap(data, stateToIdx, nStates, params?.maxGap ?? 5, params?.decay ?? 0.5);
    case 'window':
      return transitionsWindow(data, stateToIdx, nStates, params?.size ?? 3);
    case 'attention':
      return transitionsAttention(data, stateToIdx, nStates, params?.beta ?? 0.1);
    default:
      throw new Error(`Unknown transition type: ${type}`);
  }
}

function transitionsRelative(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let i = 0; i < valid.length - 1; i++) {
      const fromIdx = stateToIdx.get(valid[i]!.state);
      const toIdx = stateToIdx.get(valid[i + 1]!.state);
      if (fromIdx !== undefined && toIdx !== undefined) {
        counts.set(fromIdx, toIdx, counts.get(fromIdx, toIdx) + 1);
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

function transitionsFrequency(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let i = 0; i < valid.length - 1; i++) {
      const fromIdx = stateToIdx.get(valid[i]!.state);
      const toIdx = stateToIdx.get(valid[i + 1]!.state);
      if (fromIdx !== undefined && toIdx !== undefined) {
        counts.set(fromIdx, toIdx, counts.get(fromIdx, toIdx) + 1);
      }
    }
  }

  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights: counts, inits };
}

function transitionsCooccurrence(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let i = 0; i < valid.length - 1; i++) {
      const idx1 = stateToIdx.get(valid[i]!.state);
      const idx2 = stateToIdx.get(valid[i + 1]!.state);
      if (idx1 !== undefined && idx2 !== undefined) {
        counts.set(idx1, idx2, counts.get(idx1, idx2) + 1);
        counts.set(idx2, idx1, counts.get(idx2, idx1) + 1);
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

function transitionsReverse(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const lastIdx = stateToIdx.get(valid[valid.length - 1]!.state);
    if (lastIdx !== undefined) inits[lastIdx]!++;

    for (let i = valid.length - 1; i > 0; i--) {
      const fromIdx = stateToIdx.get(valid[i]!.state);
      const toIdx = stateToIdx.get(valid[i - 1]!.state);
      if (fromIdx !== undefined && toIdx !== undefined) {
        counts.set(fromIdx, toIdx, counts.get(fromIdx, toIdx) + 1);
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

function transitionsNgram(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
  n: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let i = 0; i <= valid.length - n; i++) {
      const fromIdx = stateToIdx.get(valid[i]!.state);
      const toIdx = stateToIdx.get(valid[i + n - 1]!.state);
      if (fromIdx !== undefined && toIdx !== undefined) {
        counts.set(fromIdx, toIdx, counts.get(fromIdx, toIdx) + 1);
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

function transitionsGap(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
  maxGap: number,
  decay: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let i = 0; i < valid.length; i++) {
      const fromIdx = stateToIdx.get(valid[i]!.state);
      if (fromIdx === undefined) continue;
      for (let j = i + 1; j < Math.min(i + maxGap + 1, valid.length); j++) {
        const toIdx = stateToIdx.get(valid[j]!.state);
        if (toIdx === undefined) continue;
        const gap = j - i;
        const weight = Math.pow(decay, gap - 1);
        counts.set(fromIdx, toIdx, counts.get(fromIdx, toIdx) + weight);
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

function transitionsWindow(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
  size: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let wStart = 0; wStart <= valid.length - size; wStart++) {
      const window = valid.slice(wStart, wStart + size);
      for (let i = 0; i < window.length; i++) {
        for (let j = i + 1; j < window.length; j++) {
          const idx1 = stateToIdx.get(window[i]!.state);
          const idx2 = stateToIdx.get(window[j]!.state);
          if (idx1 !== undefined && idx2 !== undefined) {
            counts.set(idx1, idx2, counts.get(idx1, idx2) + 1);
          }
        }
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

function transitionsAttention(
  data: SequenceData,
  stateToIdx: Map<string, number>,
  nStates: number,
  beta: number,
): { weights: Matrix; inits: Float64Array } {
  const counts = Matrix.zeros(nStates, nStates);
  const inits = new Float64Array(nStates);

  for (const row of data) {
    const valid = getValidTransitions(row);
    if (valid.length === 0) continue;

    const firstIdx = stateToIdx.get(valid[0]!.state);
    if (firstIdx !== undefined) inits[firstIdx]!++;

    for (let i = 0; i < valid.length; i++) {
      const fromIdx = stateToIdx.get(valid[i]!.state);
      if (fromIdx === undefined) continue;
      for (let j = i + 1; j < valid.length; j++) {
        const toIdx = stateToIdx.get(valid[j]!.state);
        if (toIdx === undefined) continue;
        const distance = j - i;
        const weight = Math.exp(-beta * distance);
        counts.set(fromIdx, toIdx, counts.get(fromIdx, toIdx) + weight);
      }
    }
  }

  const weights = rowNormalize(counts);
  const initSum = inits.reduce((a, b) => a + b, 0);
  if (initSum > 0) {
    for (let i = 0; i < inits.length; i++) inits[i]! /= initSum;
  }

  return { weights, inits };
}

/**
 * Compute per-sequence transition counts as a 3D array.
 * Returns array of matrices of shape (nStates x nStates),
 * one per sequence, where mat[i][j] = count of i->j transitions in that sequence.
 *
 * Matches R TNA's compute_transitions function.
 */
export function computeTransitions3D(
  data: SequenceData,
  states: string[],
  type: ModelType = 'relative',
): Matrix[] {
  const nSequences = data.length;
  const nStates = states.length;
  const stateToIdx = new Map<string, number>();
  states.forEach((s, i) => stateToIdx.set(s, i));

  const trans: Matrix[] = [];
  for (let r = 0; r < nSequences; r++) {
    trans.push(Matrix.zeros(nStates, nStates));
  }

  const nCols = data.length > 0 ? data[0]!.length : 0;

  if (type === 'relative' || type === 'frequency') {
    for (let col = 0; col < nCols - 1; col++) {
      for (let row = 0; row < nSequences; row++) {
        const fromVal = data[row]![col];
        const toVal = data[row]![col + 1];
        if (isNA(fromVal) || isNA(toVal)) continue;
        const fromIdx = stateToIdx.get(fromVal!);
        const toIdx = stateToIdx.get(toVal!);
        if (fromIdx !== undefined && toIdx !== undefined) {
          trans[row]!.set(fromIdx, toIdx, trans[row]!.get(fromIdx, toIdx) + 1);
        }
      }
    }
  } else if (type === 'reverse') {
    for (let col = 0; col < nCols - 1; col++) {
      for (let row = 0; row < nSequences; row++) {
        const fromVal = data[row]![col + 1];
        const toVal = data[row]![col];
        if (isNA(fromVal) || isNA(toVal)) continue;
        const fromIdx = stateToIdx.get(fromVal!);
        const toIdx = stateToIdx.get(toVal!);
        if (fromIdx !== undefined && toIdx !== undefined) {
          trans[row]!.set(fromIdx, toIdx, trans[row]!.get(fromIdx, toIdx) + 1);
        }
      }
    }
  } else if (type === 'co-occurrence') {
    for (let i = 0; i < nCols - 1; i++) {
      for (let j = i + 1; j < nCols; j++) {
        for (let row = 0; row < nSequences; row++) {
          const fromVal = data[row]![i];
          const toVal = data[row]![j];
          if (isNA(fromVal) || isNA(toVal)) continue;
          const fi = stateToIdx.get(fromVal!);
          const ti = stateToIdx.get(toVal!);
          if (fi !== undefined && ti !== undefined) {
            trans[row]!.set(fi, ti, trans[row]!.get(fi, ti) + 1);
            trans[row]!.set(ti, fi, trans[row]!.get(ti, fi) + 1);
          }
        }
      }
    }
  }

  return trans;
}

/**
 * Compute weight matrix from array of per-sequence transition matrices.
 * Sums over sequences, then row-normalizes for 'relative' type.
 */
export function computeWeightsFrom3D(
  transitions: Matrix[],
  type: ModelType = 'relative',
  scaling?: string | string[] | null,
): Matrix {
  if (transitions.length === 0) {
    throw new Error('No transition matrices provided');
  }

  const n = transitions[0]!.rows;
  const weights = Matrix.zeros(n, n);

  // Sum over sequences
  for (const t of transitions) {
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < n; j++) {
        weights.set(i, j, weights.get(i, j) + t.get(i, j));
      }
    }
  }

  // Row normalize for relative type
  let result = type === 'relative' ? rowNormalize(weights) : weights;

  // Apply additional scaling
  if (scaling) {
    const scaled = applyScaling(result, scaling);
    result = scaled.weights;
  }

  return result;
}

/** Process an existing weight/count matrix. */
export function computeWeightsFromMatrix(
  mat: Matrix,
  type: ModelType = 'relative',
): Matrix {
  if (type === 'relative') return rowNormalize(mat);
  return mat.clone();
}
